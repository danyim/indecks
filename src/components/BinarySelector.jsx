import React from 'react'
import PropTypes from 'prop-types'
import StylePropType from 'react-style-proptype'
import styled from 'styled-components'

class BinarySelector extends React.Component {
  static propTypes = {
    leftLabel: PropTypes.string,
    rightLabel: PropTypes.string,
    selection: PropTypes.oneOf([1, 2, null]).isRequired,
    containerStyle: StylePropType,
    buttonStyle: StylePropType,
    handleClick: PropTypes.func.isRequired,
  }

  static defaultProps = {
    leftLabel: 'Option 1',
    rightLabel: 'Option 2',
    containerStyle: {},
    buttonStyle: {},
    selection: null,
  }

  static container = styled.div`
    margin: 0.5rem;
    display: flex;
    flex-flow: row nowrap;
    overflow: none;
  `

  static button = styled.button`
    flex: 1 1 50%;
    margin: 0;
    transition: .2s all;
    padding: .2rem 1rem;
    border-radius: ${props => (props.left ? '.5rem 0 0 .5rem' : '0 .5rem .5rem 0')};
    border-right-width: ${props => (props.left ? '0' : '2px')};
    border-left-width: ${props => (props.left ? '2px' : '0')};
    border-color: #125688;
    // border-color: ${props => (props.selected ? '#125688' : '#edeeed')};
    background-color: ${props => (props.selected ? '#125688' : 'none')};
    color: ${props => (props.selected ? 'white' : '#edeeed')};
    &:hover {
      color: ${props => (props.selected ? 'white' : '#125688')};
      // border-right-width: 2px;
      // border-left-width: 2px;
    }
  `

  render() {
    return (
      <BinarySelector.container style={this.props.containerStyle}>
        <BinarySelector.button
          selected={this.props.selection === 1}
          left
          onClick={() => this.props.handleClick(1)}
          style={this.props.buttonStyle}
        >
          {this.props.leftLabel}
        </BinarySelector.button>
        <BinarySelector.button
          selected={this.props.selection === 2}
          right
          onClick={() => this.props.handleClick(2)}
          style={this.props.buttonStyle}
        >
          {this.props.rightLabel}
        </BinarySelector.button>
      </BinarySelector.container>
    )
  }
}

export default BinarySelector
