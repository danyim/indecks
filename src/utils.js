export const createReducer = (initialState, handlers) => {
  return function reducer(state = initialState, action = {}) {
    if (handlers.hasOwnProperty(action.type)) {
      return handlers[action.type](state, action);
    }
    return state;
  };
};

/**
 * Given an array of consecutive numbers, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a Array of numbers
 * @return {Number}   Missing value
 */
export const findMissing = (a) => {
  let expectedNextValue;
  for(let k of a)
  {
    expectedNextValue = a[k] + 1;
    if(k <= a.length - 1 && expectedNextValue !== a[k + 1]) {
      return expectedNextValue;
    }
  }
  return null;
};

/**
 * Given an array with potentially missing values at parameter key, find the
 * next best 0-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) => {
  return findMissing(a.map(x => x[key])) || Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1);
};

