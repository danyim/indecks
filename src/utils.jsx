import React from 'react'
import { Route } from 'react-router-dom'
import _ from 'lodash'

export const createReducer = (initialState, handlers) =>
  function reducer(state = initialState, action = {}) {
    if (Object.prototype.hasOwnProperty.call(handlers, action.type)) {
      return handlers[action.type](state, action)
    }
    return state
  }

/**
 * Given an array of consecutive numbers from 1 to n, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a   Array of numbers
 * @param  {Number} min Minimum value (0 or 1)
 * @return {Number}     Missing value
 */
export const findMissing = (a, min = 1) => {
  let expected = min
  for (let k = 0; k < a.length; k++) {
    if (a[k] !== expected) {
      return expected
    }
    expected++
  }
  return null
}

/**
 * Given an array of objects, check the value of the key and return the
 * best 1-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) =>
  findMissing(a.map(x => x[key])) ||
  Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1)

/**
 * Helper function to generate routes
 * @param  {Object} route) Route object that takes the form
 *                         {
 *                           exact: boolean,
 *                           path: string,
 *                           component: Component,
 *                           routes: Routes
 *                         }
 * @return {Component}        Returns a <Route> component to plug into a <Switch>
 */
export const routeWithSubRoutes = route =>
  <Route
    key={_.uniqueId()}
    exact={route.exact || false}
    path={route.path}
    render={props =>
      // Pass the sub-routes down to keep nesting
      // <route.component {...props}>
      //   { route.children ? route.children.map(child => routeWithSubRoutes(child)) : null }
      // </route.component>
      <route.component {...props} routes={route.routes} />}
  />
