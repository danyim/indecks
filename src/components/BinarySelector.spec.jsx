import toJSON from 'enzyme-to-json'
import 'jest-styled-components' // eslint-disable-line
import { shallow, mount, setup, setupFull } from '../test/utils' // eslint-disable-line no-unused-vars
import BinarySelector from './BinarySelector'

const defaultProps = {
  leftLabel: '',
  rightLabel: '',
  selection: 1,
  containerStyle: null,
  buttonStyle: null,
  handleClick: () => {}
}

describe('BinarySelector', () => {
  it('should render self and subcomponents', () => {
    const { wrapper } = mount(BinarySelector, defaultProps)
    const tree = toJSON(wrapper)
    expect(tree).toMatchStyledComponentsSnapshot()
  })

  it('should correctly highlight the selected value', () => {
    let { wrapper } = setupFull(BinarySelector, {
      ...defaultProps,
      selection: 2
    })

    const button = wrapper.find('button')
    const buttonSelected = wrapper.find('button[selected=true]')
    const buttonLeft = wrapper.find('button').last()
    const buttonRight = wrapper.find('button').last()

    expect(button.length).toEqual(2)
    expect(buttonSelected.exists()).toEqual(true)
    expect(buttonSelected.html()).toEqual(buttonRight.html())

    wrapper = setupFull(BinarySelector, {
      ...defaultProps,
      selection: 1
    }).wrapper

    expect(buttonSelected.html()).toEqual(buttonLeft.html())
  })

  it('should call handleClick when any selection is clicked', () => {
    const handler = jest.fn()
    const { wrapper } = setupFull(BinarySelector, {
      ...defaultProps,
      handleClick: handler
    })

    const buttonLeft = wrapper.find('button').last()
    const buttonRight = wrapper.find('button').last()

    buttonLeft.simulate('click')
    expect(handler).toHaveBeenCalled()
    buttonRight.simulate('click')
    expect(handler).toHaveBeenCalledTimes(2)
  })
})
