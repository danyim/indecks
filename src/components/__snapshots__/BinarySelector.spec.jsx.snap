// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BinarySelector should render self and subcomponents 1`] = `
.jiFHpl {
  margin: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  overflow: none;
}

.wAXXP {
  -webkit-flex: 1 1 50%;
  -ms-flex: 1 1 50%;
  flex: 1 1 50%;
  margin: 0;
  -webkit-transition: .2s all;
  transition: .2s all;
  padding: .2rem 1rem;
  border-radius: .5rem 0 0 .5rem;
  border-right-width: 0;
  border-left-width: 2px;
  border-color: #125688;
  background-color: #125688;
  color: white;
}

.wAXXP:hover {
  color: white;
}

.elbUxh {
  -webkit-flex: 1 1 50%;
  -ms-flex: 1 1 50%;
  flex: 1 1 50%;
  margin: 0;
  -webkit-transition: .2s all;
  transition: .2s all;
  padding: .2rem 1rem;
  border-radius: 0 .5rem .5rem 0;
  border-right-width: 2px;
  border-left-width: 0;
  border-color: #125688;
  background-color: none;
  color: #edeeed;
}

.elbUxh:hover {
  color: #125688;
}

<BinarySelector
  buttonStyle={null}
  containerStyle={null}
  handleClick={[Function]}
  leftLabel=""
  rightLabel=""
  selection={1}
>
  <styled.div
    style={null}
  >
    <div
      className="sc-bdVaJa jiFHpl"
      style={null}
    >
      <styled.button
        left={true}
        onClick={[Function]}
        selected={true}
        style={null}
      >
        <button
          className="sc-bwzfXH wAXXP"
          onClick={[Function]}
          selected={true}
          style={null}
        />
      </styled.button>
      <styled.button
        onClick={[Function]}
        right={true}
        selected={false}
        style={null}
      >
        <button
          className="sc-bwzfXH elbUxh"
          onClick={[Function]}
          selected={false}
          style={null}
        />
      </styled.button>
    </div>
  </styled.div>
</BinarySelector>
`;
